### ISSUE TITLE
Inference speed m2m_100

### ISSUE THREAD

**Issue by SaricVr** [pytorch/fairseq#3100]
Hello,

I'm experimenting with the m2m_100 model (4 gpus checkpoint) and observed that on my testing machine, the translation of sentences of about 128 tokens take about 2 minutes. I'm talking about pure inference time and decoding, without taking into account the time spent loading the model and preparing the input.
I was wondering if this timing is what one would expect or if there is something I can do to speed up the process... although I doubt it.
Also, is there a way to translate sentences longer than 256 tokens? I understand that max_positions is fixed and the model has been trained with that number, so changing it at inference time could disrupt the behavior, however I just want to know if this is something feasible or not. This second question is related to #2875 

Thank you

**Comment by Jorghi12** [comment]
Hello @SaricVr. My inference time is around 2 minutes per 128 sentences on an 8 core intel processor, which is about 1 second per sentence translation. 

The easiest solution is to distill knowledge into a smaller, faster architecture. For instance, I'm currently building a mobile friendly version of the M2M model by applying knowledge distillation to train a smaller architecture (e.g. fewer transformer layers, smaller layer width). The inference time will improve by doing this (at the expense of accuracy). I've already tried torchscript & packages such as torch.utils.mobile_optimizer, but received inadequate results there.

I'm also exploring the [mobileBERT architecture](https://arxiv.org/abs/2004.02984). 
> On the natural language inference tasks of GLUE, MobileBERT achieves a GLUEscore o 77.7 (0.6 lower than BERT_BASE), and 62 ms latency on a Pixel 4 phone.

Another, more involved, solution is to port the M2M model to a framework optimized for inference latency (e.g. [MarianNMT](https://marian-nmt.github.io/)). The model size may still need to be further reduced, however.

**Comment by Jorghi12** [comment]
Hello @SaricVr. My inference time is around 2 minutes per 128 sentences on an 8 core intel processor, which is about 1 second per sentence translation. 

The easiest solution is to distill knowledge into a smaller, faster architecture. 

For instance, I'm currently building a mobile friendly version of the M2M model by applying knowledge distillation to train a smaller architecture (fewer transformer layers, smaller layer width). This will improve latency (at the expense of accuracy). 
Note: I've tried torchscript & packages such as torch.utils.mobile_optimizer, but received inadequate results there.

I'm also exploring the [mobileBERT architecture](https://arxiv.org/abs/2004.02984). 
> On the natural language inference tasks of GLUE, MobileBERT achieves a GLUEscore o 77.7 (0.6 lower than BERT_BASE), and 62 ms latency on a Pixel 4 phone.

A more involved solution is to port the M2M model into a framework optimized for inference latency (e.g. [MarianNMT](https://marian-nmt.github.io/), which is a framework used by many [huggingface transformers](https://huggingface.co/transformers/model_doc/marian.html)). I suspect the model size might still need to be reduced, though.

**Comment by huihuifan** [comment]
unfortunately, the inference speed for large models is indeed very slow. We are very interested in making this faster, with alternative architectures and compression as Jorg mentions above. thanks for your interest!

**Comment by SaricVr** [comment]
Ok, understood. Thank you for the answers and for the great pointers!

**Comment by gfiameni** [comment]
Hi all,
I am also interested in reducing inference time. However, after having profiled the code, I realised most of the time is spent executing this function `fairseq/sequence_generator.py:733(_no_repeat_ngram)` which explicitly runs on CPU. According to the documentation,`for each beam and batch sentence, generate a list of previous ngrams`.

**Comment by edchengg** [comment]
> 

are you referring to the 12B checkpoint?

**Comment by edchengg** [comment]
> Hello @SaricVr. My inference time is around 2 minutes per 128 sentences on an 8 core intel processor, which is about 1 second per sentence translation.
> 
> The easiest solution is to distill knowledge into a smaller, faster architecture.
> 
> For instance, I'm currently building a mobile friendly version of the M2M model by applying knowledge distillation to train a smaller architecture (fewer transformer layers, smaller layer width). This will improve latency (at the expense of accuracy). Note: I've tried torchscript & packages such as torch.utils.mobile_optimizer, but received inadequate results there.
> 
> I'm also exploring the [mobileBERT architecture](https://arxiv.org/abs/2004.02984).
> 
> > On the natural language inference tasks of GLUE, MobileBERT achieves a GLUEscore o 77.7 (0.6 lower than BERT_BASE), and 62 ms latency on a Pixel 4 phone.
> 
> A more involved solution is to port the M2M model into a framework optimized for inference latency (e.g. [MarianNMT](https://marian-nmt.github.io/), which is a framework used by many [huggingface transformers](https://huggingface.co/transformers/model_doc/marian.html)). I suspect the model size might still need to be reduced, though.

are you referring to the 12B ckpt?

**Comment by Amy234543** [comment]
> Hello @SaricVr. My inference time is around 2 minutes per 128 sentences on an 8 core intel processor, which is about 1 second per sentence translation.
> 
> The easiest solution is to distill knowledge into a smaller, faster architecture.
> 
> For instance, I'm currently building a mobile friendly version of the M2M model by applying knowledge distillation to train a smaller architecture (fewer transformer layers, smaller layer width). This will improve latency (at the expense of accuracy). Note: I've tried torchscript & packages such as torch.utils.mobile_optimizer, but received inadequate results there.
> 
> I'm also exploring the [mobileBERT architecture](https://arxiv.org/abs/2004.02984).
> 
> > On the natural language inference tasks of GLUE, MobileBERT achieves a GLUEscore o 77.7 (0.6 lower than BERT_BASE), and 62 ms latency on a Pixel 4 phone.
> 
> A more involved solution is to port the M2M model into a framework optimized for inference latency (e.g. [MarianNMT](https://marian-nmt.github.io/), which is a framework used by many [huggingface transformers](https://huggingface.co/transformers/model_doc/marian.html)). I suspect the model size might still need to be reduced, though.
@Jorghi12  Hello, I want to ask if the knowledge distillation of m2m100 has been realized? How to choose teacher model and student model

**Comment by wenchaoliu-93** [comment]
To give a sense of how slow the inference is, I compared M2M100 418M with AWS. Keep in mind that translation via AWS involves, in addition to inference, uploading and downloading data. M2M was noticeably slower than AWS. 
The left image shows the time in seconds each function call takes, and the right image shows the code for M2M. Both were running on the same machine -- a high-end desktop Alienware.
I thought local inferencing would be much faster. How could this be? 
![Capture](https://github.com/facebookresearch/fairseq/assets/121582343/1f1359e8-6e72-47a7-af10-a56624936881)
![Capture](https://github.com/facebookresearch/fairseq/assets/121582343/686ec3c8-dcce-4711-95da-f0d9f1dc9053)

**Comment by wenchaoliu-93** [comment]
> Hello @SaricVr. My inference time is around 2 minutes per 128 sentences on an 8 core intel processor, which is about 1 second per sentence translation.
> 
> The easiest solution is to distill knowledge into a smaller, faster architecture.
> 
> For instance, I'm currently building a mobile friendly version of the M2M model by applying knowledge distillation to train a smaller architecture (fewer transformer layers, smaller layer width). This will improve latency (at the expense of accuracy). Note: I've tried torchscript & packages such as torch.utils.mobile_optimizer, but received inadequate results there.
> 
> I'm also exploring the [mobileBERT architecture](https://arxiv.org/abs/2004.02984).
> 
> > On the natural language inference tasks of GLUE, MobileBERT achieves a GLUEscore o 77.7 (0.6 lower than BERT_BASE), and 62 ms latency on a Pixel 4 phone.
> 
> A more involved solution is to port the M2M model into a framework optimized for inference latency (e.g. [MarianNMT](https://marian-nmt.github.io/), which is a framework used by many [huggingface transformers](https://huggingface.co/transformers/model_doc/marian.html)). I suspect the model size might still need to be reduced, though.

Depending on, among other things, the hardware and the model, "1 second per sentence translation" isn't that bad. See my post. That said, AWS is surprisingly faster at, often times, less than 0.1 second.
